File: tarea2-cisternas-flores/main.py
import sys

def find_maximum_subarray(arr: list):
    def find_max_crossing_subarray(arr, low, mid, high):
        left_sum = float('-inf')
        sum = 0
        max_left = 0
        for i in range(mid, low - 1, -1):
            sum += arr[i]
            if sum > left_sum:
                left_sum = sum
                max_left = i

        right_sum = float('-inf')
        sum = 0
        max_right = 0
        for i in range(mid + 1, high + 1):
            sum += arr[i]
            if sum > right_sum:
                right_sum = sum
                max_right = i

        return max_left, max_right, left_sum + right_sum

    def find_maximum_subarray_recursive(arr, low, high):
        if low == high:
            return low, high, arr[low]

        mid = (low + high) // 2

        left_low, left_high, left_sum = find_maximum_subarray_recursive(arr, low, mid)
        right_low, right_high, right_sum = find_maximum_subarray_recursive(arr, mid + 1, high)
        cross_low, cross_high, cross_sum = find_max_crossing_subarray(arr, low, mid, high)

        if left_sum >= right_sum and left_sum >= cross_sum:
            return left_low, left_high, left_sum
        elif right_sum >= left_sum and right_sum >= cross_sum:
            return right_low, right_high, right_sum
        else:
            return cross_low, cross_high, cross_sum

    if len(arr) == 0:
        return None

    low, high, max_sum = find_maximum_subarray_recursive(arr, 0, len(arr) - 1)
    return max_sum, list(range(low, high + 1))



if __name__ == "__main__":
    acciones = [int(linea.strip()) for linea in sys.stdin]

    resultado = find_maximum_subarray(acciones)

    print(str(resultado[1][0] + 1)  + " " + str(resultado[1][-1] + 1))
    print(resultado[0])

File: tarea2-cisternas-flores/main.cpp
#include <iostream>
#include <vector>
#include <limits>

struct Subarray
{
    int low;
    int high;
    int sum;
};

Subarray findMaxCrossingSubarray(const std::vector<int> &arr, int low, int mid, int high);
/**
 * @brief Encuentra el máximo crecimiento
 *
 * Busca el mayor crecimiento desde un punto intermedio en el arreglo dado.
 * No busca de forma explícita por la izquierda y derecha, por eso lo busca cruzado.
 *
 * @param arr Arreglo de enteros.
 * @param low El índice más bajo del arreglo.
 * @param mid El pivote medio del arreglo.
 * @param high El índive más alto del arreglo.
 * @return Un objeto (para inicializar el Subarray) con el nuevo máximo encontrado sus límites.
 */

Subarray findMaximumSubarrayRecursive(const std::vector<int> &arr, int low, int high);
/**
 * @brief Encuentra el máximo crecimiento de forma recursiva (para dividir -> conquistar -> combinar (al final))
 *
 * De forma recusiva esta función encuentra el máximo crecimiento para un arreglo de número dados.
 * Invocará a la función findMaxCrossingSubarray y a sí misma para tener tres comparaciones:
 * * Por la izquierda
 * * Por la derecha
 * * De forma cruzada (desde un pivote)
 * Así se conseguirá dividir de varias formas el arreglo de enteros para posteriormente hacer el proceso de "combinar"
 * y encontrar el máximo crecimiento entre todas las divisiones.
 *
 * @param arr Arreglo de enteros.
 * @param low El índice más bajo del arreglo.
 * @param high El índive más alto del arreglo.
 * @return Un objeto (para inicializar el Subarray) con el nuevo máximo encontrado sus límites.
 *
*/

Subarray findMaximumSubarray(const std::vector<int> &arr);
/**
 * @brief Encuentra el máximo crecimiento total de todo el arreglo
 *
 * Dado el arreglo de enteros (que serán leídos por entrada estándar), determinará el mayor crecimiento del mismo.
 *
 * @param arr Arreglo de enteros.
 * @return Un objeto de tipo Subarray que contendrá el máximo crecimiento y los índices de un inicio y fin del mismo.
 */

int main()
{
    std::vector<int> numbers;
    int num;
    while (std::cin >> num) numbers.push_back(num);

    Subarray result = findMaximumSubarray(numbers);

    std::cout << result.low + 1  << " " << result.high + 1 << std::endl;
    std::cout << result.sum << std::endl;

    std::cout << std::endl;

    return 0;
}

Subarray findMaxCrossingSubarray(const std::vector<int> &arr, int low, int mid, int high)
{
    int leftSum = std::numeric_limits<int>::min(); // el equivalente a hacer float('-float') en Python
    int sum = 0;
    int maxLeft = 0;
    for (int i = mid; i >= low; i--)
    {
        sum += arr[i];
        if (sum > leftSum)
        {
            leftSum = sum;
            maxLeft = i;
        }
    }

    int rightSum = std::numeric_limits<int>::min();  // el equivalente a hacer float('-float') en Python
    sum = 0;
    int maxRight = 0;
    for (int i = mid + 1; i <= high; i++)
    {
        sum += arr[i];
        if (sum > rightSum)
        {
            rightSum = sum;
            maxRight = i;
        }
    }

    return {maxLeft, maxRight, leftSum + rightSum};
}

Subarray findMaximumSubarrayRecursive(const std::vector<int> &arr, int low, int high)
{
    if (low == high) return {low, high, arr[low]};

    int mid = (low + high) / 2;

    Subarray left = findMaximumSubarrayRecursive(arr, low, mid);
    Subarray right = findMaximumSubarrayRecursive(arr, mid + 1, high);
    Subarray crossing = findMaxCrossingSubarray(arr, low, mid, high);

    if (left.sum >= right.sum && left.sum >= crossing.sum) return left;
    else if (right.sum >= left.sum && right.sum >= crossing.sum) return right;
    else return crossing;
}

Subarray findMaximumSubarray(const std::vector<int> &arr)
{
    if (arr.empty()) return {0, 0, 0};

    return findMaximumSubarrayRecursive(arr, 0, arr.size() - 1);
}
